### PR TITLE
Dallasf/concrete dynamics

### PR DESCRIPTION
**Pull request by dallasfoster**
# ALCHEMI Toolkit Pull Request

## Description

Implements concrete dynamics integrators and optimizers (NVE, NVT-Langevin, NVT-Nosé-Hoover, NPT, NPH, FIRE, FIRE2) backed by Warp GPU kernels, completes the FusedStage / inflight-batching pipeline, and introduces a    
  non-blocking observer hook mechanism.                                                                                                                                                                                      

## Type of Change

- [ x ] Bug fix (non-breaking change that fixes an issue)
- [ x ] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ x ] Performance improvement
- [ ] Documentation update
- [ ] Refactoring (no functional changes)
- [ ] CI/CD or infrastructure change

## Changes Made

Concrete integrators and optimizers
  - nvalchemi/dynamics/integrators/ — NVE (velocity-Verlet), NVTLangevin, NVTNoseHoover, NPT, NPH
  - nvalchemi/dynamics/optimizers/ — FIRE, FIRE2, variable-cell variants
  - nvalchemi/dynamics/_ops/ — low-level Warp kernel bindings for each integrator/optimizer

  FusedStage and inflight batching
  - FusedStage composition via + operator with n_steps auto-migration, init_fn, and register_fused_hook()
  - SizeAwareSampler inflight batching with monotonic system_id stamping
  - ConvergenceHook.from_forces() factory for force-norm-based stage migration
  - _prestep_sync_buffers fix: recv_buffer.zero() was not being called (bug fix)

  Data fixes
  - Batch.from_data_list used torch.stack for system-level tensors, producing wrong shapes (N, 1, D) instead of (N, D); changed to torch.cat

  Non-blocking observer hooks
  - DeferredObserverHook base class: splits hook logic into extract() (main thread, non-blocking CPU copies) and observe() (background thread, safe to call .item() / I/O); uses a CUDA event to guarantee DMA completion
  before observe() reads data
  - LoggingHook updated to use DeferredObserverHook
  - BaseDynamics.run() and FusedStage.run() call _flush_deferred_hooks() on exit so all pending log writes complete before returning

## Testing

  - examples/02_dynamics_example.py — basic NVE/NVT/FIRE usage
  - examples/03_advanced_dynamics_example.py — three-stage FIRE → NVT → NVE pipeline with inflight batching, custom hooks, and DeferredObserverHook demonstration
  - test/dynamics/test_ops.py — kernel-level tests for all integrators/optimizers
  - test/dynamics/test_inflight.py and test/dynamics/test_state_management.py — inflight batching and per-system state lifecycle

## Checklist

- [ ] I have read and understand the [Contributing Guidelines](../CONTRIBUTING.md)
- [ ] I have updated the [CHANGELOG.md](../CHANGELOG.md)
- [ ] I have performed a self-review of my code
- [ ] I have added docstrings to new functions/classes
- [ ] I have updated the documentation (if applicable)

## Additional Notes

> [!TIP]
> This repository uses Greptile, an AI code review service, to help conduct
> pull request reviews. We encourage contributors to read and consider suggestions
> made by Greptile, but note that human maintainers will provide the necessary
> reviews for merging: Greptile's comments are **not** a qualitative judgement
> of your code, nor is it an indication that the PR will be accepted/rejected.
> We encourage the use of emoji reactions to Greptile comments, depending on
> their usefulness and accuracy.

### CODE REVIEW

File: examples/02_dynamics_example.py [hunk 1]
```diff
@@ -0,0 +1,192 @@
...
+# ``ON_CONVERGE`` hooks when fmax < 0.05.  The ``FmaxLogHook`` prints
+# progress at the ``AFTER_STEP`` stage every ``frequency`` steps.
+
+print("=== Part 1: FIRE Geometry Optimization ===")
+
+data_list_opt = [_make_system(n, seed) for n, seed in [(4, 1), (6, 2), (5, 3)]]
+batch_opt = Batch.from_data_list(data_list_opt)
+print(f"Batch: {batch_opt.num_graphs} systems, {batch_opt.num_nodes} atoms total\n")
+
+from nvalchemi.dynamics.base import ConvergenceHook
```
Comment (laserkelvin) [inline examples/02_dynamics_example.py:105]: I think this is fine, but I also intended it so that users could also just pass a dict directly in and not *need* to know and import it.

As an example I think we could either show the explicit import or not

File: examples/02_dynamics_example.py [hunk 2]
```diff
@@ -0,0 +1,192 @@
...
+    # ConvergenceHook evaluates per-atom force norms, scatter-maxes them to
+    # per-system fmax, and marks systems with fmax <= 0.05 as converged.
+    convergence_hook=ConvergenceHook(
+        criteria=[
+            {"key": "forces", "threshold": 0.05, "reduce_op": "norm", "reduce_dims": -1}
+        ]
+    ),
+)
+
+batch_opt = fire_opt.run(batch_opt)
```
Comment (laserkelvin) [inline examples/02_dynamics_example.py:120]: Did we want to have `fire_opt(batch_opt)`?

File: examples/02_dynamics_example.py [hunk 3]
```diff
@@ -0,0 +1,192 @@
...
+
+print("\n\n=== Part 2: FusedStage — FIRE + NVTLangevin ===")
+
+
+# Build a fresh batch and attach status.
+data_list_fused = [_make_system(n, seed) for n, seed in [(4, 10), (6, 11), (5, 12)]]
+batch_fused = Batch.from_data_list(data_list_fused)
+
+# All systems start in the FIRE stage (status = 0).
+batch_fused["status"] = torch.zeros(batch_fused.num_graphs, 1, dtype=torch.long)
```
Comment (laserkelvin) [inline examples/02_dynamics_example.py:151]: Should we make it so that the user doesn't have to initialize this tensor?

I think the sharp edge is that it's hidden within `Batch`, instead of e.g. passing it into `dynamics.run` as a required argument

File: examples/03_advanced_dynamics_example.py [hunk 4]
```diff
@@ -0,0 +1,880 @@
...
+#   * ``extract()``  — runs in the main thread; copies tensors to CPU
+#     with ``non_blocking=True``.
+#   * ``observe()``  — runs in a background thread; safe to call
+#     ``.item()``, accumulate Python state, and print.
+#
+# ``stage = HookStageEnum.AFTER_COMPUTE`` is set as a class attribute
+# to override the ``AFTER_STEP`` default from ``DeferredObserverHook``.
+
+
+class EnergyTracker(DeferredObserverHook):
```
Comment (laserkelvin) [inline examples/03_advanced_dynamics_example.py]: Should update this depending on what we decide to do with the `DeferredObserverHook`

File: examples/03_advanced_dynamics_example.py [hunk 5]
```diff
@@ -0,0 +1,880 @@
...
+# ---- Stage 0: FIRE ----
+# ConvergenceHook.from_forces() drives both standalone FIRE convergence detection
+# AND the auto-registered 0→1 migration hook that FusedStage creates.
+# FusedStage inherits the criteria from fire_stage.convergence_hook, so the
+# same force-norm threshold governs the inter-stage migration.
+fire_stage = FIRE(
+    model=model,
+    dt=0.1,
+    convergence_hook=ConvergenceHook.from_forces(FIRE_FMAX_THRESHOLD),
+    device_type="cuda:0",
```
Comment (laserkelvin) [inline examples/03_advanced_dynamics_example.py:333]: `device_type` should just be `cuda`, without the index

File: examples/03_advanced_dynamics_example.py [hunk 6]
```diff
@@ -0,0 +1,880 @@
...
+    dt=0.5,
+    n_steps=PROD_STEPS_PER_SYSTEM,
+    device_type="cuda:0",
+)
+
+# --- Register custom hooks on the sub-stages BEFORE fusing ---
+
+# EnergyTracker on the FIRE sub-stage: print energy every 10 steps
+fire_energy_tracker = EnergyTracker(frequency=10, label="FIRE-energy")
+fire_stage.register_hook(fire_energy_tracker)
```
Comment (laserkelvin) [inline examples/03_advanced_dynamics_example.py]: These should be passed to the constructors as `hooks=[...]`, instead of calling register hook directly

File: pyproject.toml [hunk 7]
```diff
@@ -74,7 +74,7 @@ override-dependencies = [
 
 [tool.uv.sources]
 nvidia-physicsnemo = { git = "https://github.com/NVIDIA/physicsnemo", rev = "70b06ed" }
-nvalchemi-toolkit-ops = { git = "https://github.com/NVIDIA/nvalchemi-toolkit-ops", rev = "82f2e70" }
+nvalchemi-toolkit-ops = { git = "https://github.com/dallasfoster/nvalchemi-toolkit-ops.git", branch = "dallasf/bug_fixes" }
```
Comment (laserkelvin) [inline pyproject.toml:77]: I think this is merged in, so pin to `0.3.0-rc`?

File: examples/02_dynamics_example.py
Comment (laserkelvin) [inline examples/02_dynamics_example.py:1]: This script doesn't give that much feedback to the user - when run it just shows that it "finished X steps" for each integrator which I guess isn't too satisfying. I think just showing initial and final states is good

File: nvalchemi/data/batch.py [hunk 8]
```diff
@@ -57,8 +56,11 @@
 _EXCLUDED_KEYS = frozenset({"batch", "ptr", "device", "dtype", "info"})
 
 
-class Batch(BaseModel, DataMixin):
-    """Graph-aware Pydantic batch built on :class:`MultiLevelStorage`.
+_OWN_ATTRS = frozenset({"device", "keys", "_storage", "_data_class"})
+
+
+class Batch(DataMixin):
```
Comment (laserkelvin) [inline nvalchemi/data/batch.py:62]: I'm okay with not inheriting `BaseModel`, but was there a motivation for this?

File: nvalchemi/data/batch.py [hunk 9]
```diff
@@ -71,30 +73,37 @@ class Batch(BaseModel, DataMixin):
...
-    model_config = ConfigDict(
-        arbitrary_types_allowed=True,
-        validate_assignment=True,
-    )
-
     def __init__(
-        self, *, storage: MultiLevelStorage | None = None, **kwargs: Any
+        self,
+        *,
+        device: torch.device | str,
```
Comment (laserkelvin) [inline nvalchemi/data/batch.py:85]: `str` was dropped from the docstring

File: nvalchemi/data/batch.py [hunk 10]
```diff
@@ -71,30 +73,37 @@ class Batch(BaseModel, DataMixin):
...
+        storage: MultiLevelStorage | None = None,
+        keys: dict[str, set[str]] | None = None,
     ) -> None:
-        super().__init__(**kwargs)
-        object.__setattr__(self, "_data_class", AtomicData)
         object.__setattr__(
```
Comment (laserkelvin) [inline nvalchemi/data/batch.py:89]: I feel like this was needed to circumvent stuff when we inherited from `BaseModel`, and it should be fine to use `setattr(...)` now instead, right?

File: nvalchemi/dynamics/_ops/_bridge.py
Comment (laserkelvin) [inline nvalchemi/dynamics/_ops/_bridge.py:1]: I'm somewhat surprised we needed these functions at all - I thought all the casting and whatnot would be done at the ops level

File: nvalchemi/data/batch.py [hunk 11]
```diff
@@ -583,7 +580,7 @@ def get_data(self, idx: int) -> AtomicData:
         system = self._system_group
         if system is not None:
             for key, tensor in system.items():
-                data[key] = tensor[idx]
+                data[key] = tensor[idx].unsqueeze(0)
```
Comment (laserkelvin) [inline nvalchemi/data/batch.py:583]: This feels somewhat unexpected - how come we have to do shape management here?

File: nvalchemi/dynamics/integrators/nph.py [hunk 12]
```diff
@@ -0,0 +1,294 @@
...
+                # Pre-allocated scratch tensors; zeroed by the kernel each call.
+                "kinetic_tensors": torch.zeros(M, 9, dtype=dtype, device=dev),
+                "pressure_tensors": torch.zeros(M, 9, dtype=dtype, device=dev),
+                "volumes": torch.zeros(M, dtype=dtype, device=dev),
+                "kinetic_energy": torch.zeros(M, dtype=dtype, device=dev),
+            },
+            dev,
+        )
+
+    def _make_new_state(self, n: int, template_batch: Batch) -> Batch:
```
Comment (laserkelvin) [inline nvalchemi/dynamics/integrators/nph.py:156]: Private, but could do with some docstrings for ease of maintainence

File: nvalchemi/dynamics/optimizers/fire.py [hunk 13]
```diff
@@ -0,0 +1,527 @@
...
+            return (
+                self._uphill_init.expand(M)
+                .to(dtype=torch.int32, device=dev)
+                .contiguous()
+            )
+        val = 1 if self._uphill_init else 0
+        return torch.full((M,), val, dtype=torch.int32, device=dev)
+
+    def _init_state(self, batch: Batch) -> None:
+        M = batch.num_graphs
```
Comment (laserkelvin) [inline nvalchemi/dynamics/optimizers/fire.py:178]: Short variable name, as well as consistency across the other implementations

File: nvalchemi/dynamics/optimizers/fire2.py [hunk 14]
```diff
@@ -0,0 +1,366 @@
...
+from nvalchemi.dynamics._ops.npt_nph import stress_to_cell_force
+from nvalchemi.dynamics.base import BaseDynamics
+
+if TYPE_CHECKING:
+    from nvalchemi.dynamics.base import ConvergenceHook, Hook
+    from nvalchemi.models.base import BaseModelMixin
+
+__all__ = ["FIRE2", "FIRE2VariableCell"]
+
+_FIRE2_DEFAULTS = dict(
```
Comment (laserkelvin) [inline nvalchemi/dynamics/optimizers/fire2.py:62]: Why this pattern, and not just set the defaults to the class constructor?

I could understand if you wanted a configuration data structure to collect these; i.e. a pydantic model with some validation

File: nvalchemi/dynamics/optimizers/fire2.py [hunk 15]
```diff
@@ -0,0 +1,366 @@
...
+            tmax=self.tmax,
+            tmin=self.tmin,
+            maxstep=self.maxstep,
+        )
+
+    def post_update(self, batch: Batch) -> None:
+        """No-op; forces from new positions are used on the next step."""
+
+
+class FIRE2VariableCell(BaseDynamics):
```
Comment (laserkelvin) [inline nvalchemi/dynamics/optimizers/fire2.py:228]: This inherits from `BaseDynamics` instead of the first `FIRE2` - is this okay?

File: nvalchemi/dynamics/base.py [hunk 16]
```diff
@@ -1261,11 +1237,41 @@ class BaseDynamics(_CommunicationMixin):
     __needs_keys__: set[str] = set()
     __provides_keys__: set[str] = set()
 
+    _mutable_fields: tuple[str, ...] = ("positions", "velocities")
+
+    _bookkeeping_keys: dict[str, Callable[[int, torch.device], torch.Tensor]] = {
```
Comment (laserkelvin) [inline nvalchemi/dynamics/base.py:1253]: Could you add this to the docstring, and explain what and how this attribute is used?

File: nvalchemi/dynamics/base.py [hunk 17]
```diff
@@ -1372,6 +1387,22 @@ def _call_hooks(self, stage: HookStageEnum, batch: Batch) -> None:
             if self.step_count % hook.frequency == 0:
                 hook(batch, self)
 
+    def _flush_deferred_hooks(self) -> None:
```
Comment (laserkelvin) [inline nvalchemi/dynamics/base.py]: We can revisit this with the snapshot hook

File: nvalchemi/dynamics/base.py [hunk 18]
```diff
@@ -1649,17 +1779,17 @@ def run(self, batch: Batch, n_steps: int | None = None) -> Batch:
             )
         for _ in range(resolved):
             batch, _converged = self.step(batch)
+        self._flush_deferred_hooks()
         return batch
 
-    def _refill_check(self, batch: Batch, exit_status: int) -> Batch | None:
+    def refill_check(self, batch: Batch, exit_status: int) -> Batch | None:
```
Comment (laserkelvin) [inline nvalchemi/dynamics/base.py:1839]: Why go public?

File: nvalchemi/dynamics/sampler.py [hunk 19]
```diff
@@ -386,9 +406,106 @@ def exhausted(self) -> bool:
...
+    def request_replacements(
+        self,
+        node_counts: torch.Tensor,
+        edge_counts: torch.Tensor,
+    ) -> list[AtomicData | None]:
+        """Request replacement samples for multiple graduated systems using GPU-native constraint checking.
+
+        Eliminates the ``.tolist()`` D→H syncs from ``_refill_check``. Constraint
+        checking is fully vectorized on GPU. M scalar ``item()`` calls remain
+        (unavoidable: Python dataset indexing requires CPU integers).
```
Comment (laserkelvin) [inline nvalchemi/dynamics/sampler.py:450]: Maybe not unavoidable, but we could make a note here potentially to refactor `Dataset`
